### ISSUE TITLE
IOError: [Errno 2] No such file or directory: 'data/*/reconstruction.json'

### ISSUE THREAD

**Issue by ZoranPavlovic** [mapillary/OpenSfM#76]
Using the provided Dockerfile, a docker image was created. I then followed the instructions from the main page.

> root@7f41108d41c4:/source/OpenSfM# bin/run_all data/DS_2/
> 2016-04-17 13:37:57,367 Matching 0 image pairs
> 2016-04-17 13:37:57,800 reading features
> 2016-04-17 13:37:57,800 Merging features onto tracks
> 2016-04-17 13:37:57,800 Good tracks: 0
> images 0
> nonfisheye images 0
> 0 partial reconstructions in total.
> Traceback (most recent call last):
>   File "/source/OpenSfM/bin/opensfm", line 34, in <module>
>     command.run(args)
>   File "/source/OpenSfM/opensfm/commands/mesh.py", line 22, in run
>     reconstructions = data.load_reconstruction()
>   File "/source/OpenSfM/opensfm/dataset.py", line 289, in load_reconstruction
>     with open(self.__reconstruction_file(filename)) as fin:
> IOError: [Errno 2] No such file or directory: 'data/DS_2/reconstruction.json'

**Comment by ZoranPavlovic** [comment]
Used the config.yaml from the berlin dataset instead, and got further along:

> root@ee0fa096d028:/source/OpenSfM# bin/run_all data/DS_2/
> 2016-04-17 14:00:33,152 Extracting focal lengths for image IMG_20160417_150128.jpg
> 2016-04-17 14:00:33,167 Extracting focal lengths for image IMG_20160417_150124.jpg
> 2016-04-17 14:00:33,176 Extracting focal lengths for image IMG_20160417_150136.jpg
> 2016-04-17 14:00:33,633 Extracting ROOT_HAHOG features for image IMG_20160417_150128.jpg
> 2016-04-17 14:00:33,634 Extracting ROOT_HAHOG features for image IMG_20160417_150136.jpg
> 2016-04-17 14:00:33,633 Extracting ROOT_HAHOG features for image IMG_20160417_150124.jpg
> 2016-04-17 14:00:37,739 Found 5027 points in 3.67584204674s
> 2016-04-17 14:00:38,023 Found 5159 points in 3.90138602257s
> 2016-04-17 14:00:38,286 Found 5632 points in 3.94010305405s
> 2016-04-17 14:00:39,709 Matching 3 image pairs
> 2016-04-17 14:00:39,736 Matching IMG_20160417_150124.jpg  -  1 / 3
> 2016-04-17 14:00:39,740 Matching IMG_20160417_150136.jpg  -  2 / 3
> 2016-04-17 14:00:39,742 Matching IMG_20160417_150128.jpg  -  3 / 3
> 2016-04-17 14:00:41,234 IMG_20160417_150128.jpg - IMG_20160417_150136.jpg has 201 candidate matches
> 2016-04-17 14:00:41,234 Robust matching time : 0.000558853149414s
> 2016-04-17 14:00:41,234 Full matching 184 / 201, time: 1.49267697334s
> 2016-04-17 14:00:41,308 IMG_20160417_150124.jpg - IMG_20160417_150128.jpg has 222 candidate matches
> 2016-04-17 14:00:41,309 Robust matching time : 0.000576972961426s
> 2016-04-17 14:00:41,310 Full matching 213 / 222, time: 1.57349801064s
> 2016-04-17 14:00:42,728 IMG_20160417_150124.jpg - IMG_20160417_150136.jpg has 61 candidate matches
> 2016-04-17 14:00:42,728 Robust matching time : 0.000245094299316s
> 2016-04-17 14:00:42,728 Full matching 59 / 61, time: 1.41863298416s
> 2016-04-17 14:00:43,231 reading features
> 2016-04-17 14:00:43,260 Merging features onto tracks
> 2016-04-17 14:00:43,264 Good tracks: 391
> images 3
> nonfisheye images 3
> 0 partial reconstructions in total.
> Traceback (most recent call last):
>   File "/source/OpenSfM/bin/opensfm", line 34, in <module>
>     command.run(args)
>   File "/source/OpenSfM/opensfm/commands/mesh.py", line 22, in run
>     reconstructions = data.load_reconstruction()
>   File "/source/OpenSfM/opensfm/dataset.py", line 289, in load_reconstruction
>     with open(self.__reconstruction_file(filename)) as fin:
> IOError: [Errno 2] No such file or directory: 'data/DS_2/reconstruction.json'

**Comment by miguelriemoliveira** [comment]
Hello I am having the same problem exactly using a dataset of my own (3 images)

**Comment by paulinus** [comment]
It looks like the algorithm is unable to start a reconstruction. This may happen for different reasons such as the scene being flat, or the lack of parallax between the images.
Could you share the images here?

**Comment by BorgesJVT** [comment]
Hi!
I am having this same problem:

2017-02-09 03:55:02,799 Matching 0 image pairs
2017-02-09 03:55:03,173 reading features
2017-02-09 03:55:03,173 Merging features onto tracks
2017-02-09 03:55:03,174 Good tracks: 0
2017-02-09 03:55:03,553 Starting incremental reconstruction
2017-02-09 03:55:03,553 0 partial reconstructions in total.
Traceback (most recent call last):
  File "/home/borgesjvt/Downloads/OpenSfM-master/bin/opensfm", line 34, in <module>
    command.run(args)
  File "/home/borgesjvt/Downloads/OpenSfM-master/opensfm/commands/mesh.py", line 22, in run
    reconstructions = data.load_reconstruction()
  File "/home/borgesjvt/Downloads/OpenSfM-master/opensfm/dataset.py", line 325, in load_reconstruction
    with open(self.__reconstruction_file(filename)) as fin:
IOError: [Errno 2] No such file or directory: 'data/lund/images/reconstruction.json'

As you can see, i am trying to run using the examples datasets, lund and berlim.
Any clue about  why that is not working?

**Comment by paulinus** [comment]
@BorgesJVT there seems to be something wrong with your installation or the command you use to start the reconstruction. Did you manage to solve it?

**Comment by angussmitchell** [comment]
I have the same issue, providing my own images. It seems as if the problem is with matching image pairs, which ultinately make the creation of a reconstruction.jon impossible:

 017-09-07 17:17:52,092 Matching 0 image pairs
2017-09-07 17:17:52,136 Matching 281115_2772.JPG  -  1 / 10
2017-09-07 17:17:52,136 Matching 281115_2770.JPG  -  2 / 10
2017-09-07 17:17:52,136 Matching 281115_2771.JPG  -  3 / 10
2017-09-07 17:17:52,137 Matching 281115_2767.JPG  -  5 / 10
2017-09-07 17:17:52,137 Matching 281115_2768.JPG  -  4 / 10
2017-09-07 17:17:52,138 Matching 281115_2774.JPG  -  7 / 10
2017-09-07 17:17:52,138 Matching 281115_2766.JPG  -  8 / 10
2017-09-07 17:17:52,138 Matching 281115_2773.JPG  -  9 / 10
2017-09-07 17:17:52,139 Matching 281115_2765.JPG  -  6 / 10
2017-09-07 17:17:52,139 Matching 281115_2769.JPG  -  10 / 10
2017-09-07 17:17:52,750 reading features
2017-09-07 17:17:52,947 Merging features onto tracks
2017-09-07 17:17:52,947 Good tracks: 0
2017-09-07 17:17:53,418 Starting incremental reconstruction
2017-09-07 17:17:53,456 0 partial reconstructions in total.
Traceback (most recent call last):
  File "/home/a/geoInteractive/OpenSfM/bin/opensfm", line 34, in <module>
    command.run(args)
  File "/home/a/geoInteractive/OpenSfM/opensfm/commands/mesh.py", line 22, in run
    reconstructions = data.load_reconstruction()
  File "/home/a/geoInteractive/OpenSfM/opensfm/dataset.py", line 355, in load_reconstruction
    with open(self.__reconstruction_file(filename)) as fin:
IOError: [Errno 2] No such file or directory: 'data/sewer/reconstruction.json'

I have tried increasing the minimum features threshold, decreasing the 'lowes_ratio' and changing the feature extractor between SIFT SURF and HAHOG. The fact that no features are matched seems strange to me, seeing as my images are of equal resolution to the berlin example set and contain good overlap and camera position change. 

My own implementations of features extraction and matching using opencv's SIFT and SURF features has proved that it is possible to match features between these images. I will soon attempt to use some example visual odometry datasets from the web, to verify that features are not being matched correctly for non example datasets.

**Comment by angussmitchell** [comment]
Following up on my previous comment, I was able to get some reconstruction working. To do this, I needed to change the values of the preemptive matching options. Debugging the match_features command in an ide I found that the settings matching_time_neighbors forced the matching command to not find any matches for any image pairs (I dont have GPS measurements for my data yet, but I assume I'll need to set matching_gps_neigbors when i have GPS data.) Simply by setting matching_time_neighbors to 1, I start getting a reconstruction. This is curious, because in the master config.yaml, it states "# Number of images to match selected by time taken. Set to 0 to use no limit". This is not true, as setting it to 0 results in no preemtive matching being performed. 

I hope this helps others in the future. 

TLDR: put this in the dataset's config.yaml to overide settings:


#preemtive matching options
matching_gps_distance: 150            # Maximum gps distance between two images for matching
matching_gps_neighbors: 1             # Number of images to match selected by GPS distance. Set to 0 to use no limit
matching_time_neighbors: 1           # Number of images to match selected by time taken. Set to 0 to use no limit
preemptive_max: 200                   # Number of features to use for preemptive matching
preemptive_threshold: 1               # If number of matches passes the threshold -> full feature matching

**Comment by YanNoun** [comment]
`preemptive_threshold` is gone, so it shouldn't cause more trouble.

Regarding the non-existing reconstruction, it is the case when there's no matches and/or there's not enough parallax between any pair.
